File: queues/servicebus/Go/servicebusclient/servicebusclient.go
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Azure Service Bus helpers for Core Auto. Send from step scripts; Receive for ingress bridges only.

package servicebusclient

import (
	"context"
	"encoding/json"
	"os"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
	"github.com/core-df/coreauto-scripts-pub/queues/servicebus/Go/internal/result"
)

func connectionString() string {
	return os.Getenv("SERVICE_BUS_CONNECTION_STRING")
}

func queueName(explicit string) string {
	if explicit != "" {
		return explicit
	}
	return os.Getenv("SERVICE_BUS_QUEUE_NAME")
}

func encode(value any) []byte {
	switch v := value.(type) {
	case []byte:
		return v
	case string:
		return []byte(v)
	default:
		b, _ := json.Marshal(value)
		return b
	}
}

func decode(raw []byte) any {
	var body any
	if err := json.Unmarshal(raw, &body); err == nil {
		return body
	}
	return string(raw)
}

// Init verifies Service Bus connection settings.
func Init() result.Result {
	if connectionString() == "" {
		return result.MissingEnv("SERVICE_BUS_CONNECTION_STRING")
	}
	if os.Getenv("SERVICE_BUS_QUEUE_NAME") == "" {
		return result.MissingEnv("SERVICE_BUS_QUEUE_NAME (or pass queue per call)")
	}
	return result.Result{StatusCode: 200}
}

// Send publishes a message to a Service Bus queue.
func Send(value any, queue string) result.Result {
	conn := connectionString()
	q := queueName(queue)
	if conn == "" {
		return result.MissingEnv("SERVICE_BUS_CONNECTION_STRING")
	}
	if q == "" {
		return result.MissingEnv("SERVICE_BUS_QUEUE_NAME")
	}
	ctx := context.Background()
	client, err := azservicebus.NewClientFromConnectionString(conn, nil)
	if err != nil {
		return result.TransportError(err.Error())
	}
	sender, err := client.NewSender(q, nil)
	if err != nil {
		return result.TransportError(err.Error())
	}
	defer sender.Close(ctx)

	if err := sender.SendMessage(ctx, &azservicebus.Message{Body: encode(value)}, nil); err != nil {
		return result.TransportError(err.Error())
	}
	return result.Result{StatusCode: 200}
}

// Receive receives messages from a Service Bus queue (ingress bridges only).
func Receive(queue string, timeoutSec float64, maxMessages int, complete bool) result.Result {
	conn := connectionString()
	q := queueName(queue)
	if conn == "" {
		return result.MissingEnv("SERVICE_BUS_CONNECTION_STRING")
	}
	if q == "" {
		return result.MissingEnv("SERVICE_BUS_QUEUE_NAME")
	}
	if maxMessages < 1 {
		maxMessages = 1
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutSec*float64(time.Second)))
	defer cancel()

	client, err := azservicebus.NewClientFromConnectionString(conn, nil)
	if err != nil {
		return result.TransportError(err.Error())
	}
	receiver, err := client.NewReceiverForQueue(q, nil)
	if err != nil {
		return result.TransportError(err.Error())
	}
	defer receiver.Close(ctx)

	batch, err := receiver.ReceiveMessages(ctx, maxMessages, nil)
	if err != nil {
		return result.TransportError(err.Error())
	}
	messages := make([]map[string]any, 0, len(batch))
	for _, msg := range batch {
		raw := msg.Body
		messages = append(messages, map[string]any{
			"queue":      q,
			"message_id": msg.MessageID,
			"value":      decode(raw),
		})
		if complete {
			if err := receiver.CompleteMessage(ctx, msg, nil); err != nil {
				return result.TransportError(err.Error())
			}
		}
	}
	return result.Result{StatusCode: 200, Messages: messages}
}

File: queues/servicebus/Kotlin/src/main/kotlin/com/coredf/servicebus/ServiceBusClient.kt
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.coredf.servicebus

import com.azure.messaging.servicebus.*
import java.time.Duration

object ServiceBusClient {

    private fun env(key: String, fallback: String = ""): String = (System.getenv(key) ?: "").ifEmpty { fallback }
    private fun envOr(key: String, fallback: String): String = env(key).ifEmpty { fallback }

    private fun encode(value: Any?): ByteArray = when (value) {
        null -> ByteArray(0)
        is ByteArray -> value
        is String -> value.toByteArray(java.nio.charset.StandardCharsets.UTF_8)
        else -> JsonUtil.stringify(value).toByteArray(java.nio.charset.StandardCharsets.UTF_8)
    }

    private fun decode(raw: ByteArray?): Any? {
        if (raw == null) return null
        return try {
            JsonUtil.parse(String(raw, java.nio.charset.StandardCharsets.UTF_8))
        } catch (_: Exception) {
            String(raw, java.nio.charset.StandardCharsets.UTF_8)
        }
    }

    @JvmStatic
    fun Init(): Result {
        if (env("SERVICE_BUS_CONNECTION_STRING").isEmpty()) return Result.missingEnv("SERVICE_BUS_CONNECTION_STRING")
        if (env("SERVICE_BUS_QUEUE_NAME").isEmpty()) return Result.missingEnv("SERVICE_BUS_QUEUE_NAME (or pass queue per call)")
        return Result.ok()
    }

    @JvmStatic
    fun Send(value: Any?, queue: String? = null): Result {
        val conn = env("SERVICE_BUS_CONNECTION_STRING")
        val q = queue ?: env("SERVICE_BUS_QUEUE_NAME")
        if (conn.isEmpty()) return Result.missingEnv("SERVICE_BUS_CONNECTION_STRING")
        if (q.isEmpty()) return Result.missingEnv("SERVICE_BUS_QUEUE_NAME")
        return try {
            ServiceBusClientBuilder().connectionString(conn).sender().queueName(q).buildClient().use { sender ->
                sender.sendMessage(ServiceBusMessage(encode(value)))
            }
            Result.ok()
        } catch (e: Exception) { Result.transportError(e.message) }
    }

    @JvmStatic
    fun Receive(queue: String? = null, timeoutSec: Double = 30.0, maxMessages: Int = 1, complete: Boolean = true): Result {
        val conn = env("SERVICE_BUS_CONNECTION_STRING")
        val q = queue ?: env("SERVICE_BUS_QUEUE_NAME")
        if (conn.isEmpty()) return Result.missingEnv("SERVICE_BUS_CONNECTION_STRING")
        if (q.isEmpty()) return Result.missingEnv("SERVICE_BUS_QUEUE_NAME")
        val messages = mutableListOf<Map<String, Any?>>()
        return try {
            ServiceBusClientBuilder().connectionString(conn).receiver().queueName(q).buildClient().use { receiver ->
                for (msg in receiver.receiveMessages(maxMessages, Duration.ofSeconds(timeoutSec.toLong()))) {
                    messages.add(mapOf("queue" to q, "message_id" to msg.messageId, "value" to decode(msg.body.toBytes())))
                    if (complete) receiver.complete(msg)
                }
            }
            Result.ok(mapOf("messages" to messages))
        } catch (e: Exception) { Result.transportError(e.message) }
    }
}
